model/src/main/kotlin/utils/FindingsMatcher.kt

```diff
@@ -127,68 +111,101 @@ class FindingsMatcher(
 
-            matchedFindings.forEach { (license, copyrightFindings) ->
-                copyrightsByLicense.getOrPut(license) { mutableSetOf() } += copyrightFindings
-            }
+            matchedFindings += matchedFileFindings
+            unmatchedCopyrights += copyrights.toSet() - matchedFindings.values.flatten()
+        }
 
-            // Associate all unmatched copyright findings with all root licenses.
-            val unmatchedCopyrights = copyrights.toSet() - matchedFindings.values.flatten()
-            rootLicenses.forEach { license ->
-                copyrightsByLicense.getOrPut(license) { mutableSetOf() } += unmatchedCopyrights
+        val rootLicenseFindings = getRootLicenseFindings(licenseFindings)
+
+        return if (rootLicenseFindings.isEmpty()) {
+            FindingsMatcherResult(matchedFindings, unmatchedCopyrights)
+        } else {
+            // Associate all unmatched copyright findings with all root licenses findings.
+            rootLicenseFindings.forEach { licenseFinding ->
+                matchedFindings[licenseFinding] = matchedFindings.getValue(licenseFinding) + unmatchedCopyrights
             }
+
+            FindingsMatcherResult(matchedFindings, emptySet())
         }
+    }
 
-        return (copyrightsByLicense.keys + locationsByLicense.keys).mapTo(mutableSetOf()) { license ->
-            LicenseFindings(
-                license,
-                locationsByLicense[license] ?: sortedSetOf(),
-                copyrightsByLicense[license].orEmpty().toCopyrightFindings().toSortedSet()
-            )
+    /**
+     * Return an association of the given [copyrightFindings] to [licenseFindings].
+     * Copyright findings are either matched to a license finding located nearby in the same file or to a license
+     * finding pointing to a license file. Whether a file is a license file is determined by the
+     * [FileMatcher] passed to the constructor. All [CopyrightFindings]s which cannot be matched are not present
+     * in the result while all given [licenseFindings] are contained in the result exactly once.
+     */
+    fun match(licenseFindings: Collection<LicenseFinding>, copyrightFindings: Collection<CopyrightFinding>):
+            Set<LicenseFindings> {
+        val result = matchFindings(licenseFindings.toSet(), copyrightFindings.toSet())
+
+        return result.matchedFindings.entries.groupBy { it.key.license }.mapTo(mutableSetOf()) { (license, findings) ->
+            val locations = findings.mapTo(sortedSetOf()) { it.key.location }
+
+            val copyrights = findings.flatMap { it.value }.groupBy { it.statement }
+                .mapTo(sortedSetOf()) { (statement, findings) ->
+                    CopyrightFindings(statement, findings.mapTo(sortedSetOf()) { it.location })
+                }
+
+            LicenseFindings(license, locations, copyrights)
         }
     }
 }
+
+/**
+ * The result of the [FindingsMatcher].
+ */
+data class FindingsMatcherResult(
```

Comment:
out of scope: It might be interesting to think about how `root license` things could be communicated via this (and the new) license API. e.g. how it'd be exposed and where the associations of copyrights to root licenses would take place.

That could e.g. be valuable for the notice by package reporter to determine the contents of the actual license files.

Reply:
Good point, I'll keept that in mind while working on the license API, and probably make another change here if required. I think there are also other things that could define a root license, e.g. declared licenses could be root licenses as well.

Reply:
@mnonnenmacher you may wanna note also #2300.